contracts/NFcharT.sol

```diff
@@ -69,40 +79,358 @@ contract NFcharT is ERC721Enumerable, Ownable, ReentrancyGuard {
+        toReturn = abi.encodePacked(
+            toReturn,
+            remainingPointsSVG(twips, normalizedTwaps, xInterval)
+        );
+
+        return toReturn;
+    }
+
+    /*
+    * @dev Returns an array of uints of y coords
+    * representing price data that is normalized
+    * to fit a 250px square chart.
+    */
+    function calcNormalizedTwaps(uint256[] memory twips) internal pure returns (uint256[7] memory normalizedTwaps) {
+        // TODO test for inactive markets or invalid values
+        //      ... what if the price oracle returns 0 or -1?
+        // thanks to Ross Bulat for normalizing the data for SVG
+        // https://tinyurl.com/563uk68a
+
+        // TODO format twips to decimal places
+        // get min and max of price data
+        uint256 min = twips[0];
+        uint256 max = twips[0];
+        for (uint8 i=1; i<twips.length; i++) {
+            if (twips[i] > max) max = twips[i];
+            if (twips[i] < min) min = twips[i];
+        }
+
+        for (uint8 i=0; i<twips.length; i++) {
+            // feature scaling, multiplied by 250 to fit the 250px square graph
+
+            // feature scaling gives a result between 0 and 1; in solidity this
+            // will round down to 0, so the result is multiplied and later divided
+            //  by the max to preserve the precision
+
+            // since the chart is a square graph that lies between 50px and 300px,
+            // subtract the result from 300 to inverse the coords for SVG
+
+            // a 3px buffer is given for the circles on the graph,
+            // so the first coord starts at 297 instead of 300
+            normalizedTwaps[i] = 297 - ( ( ( (twips[i] - min) * max ) / (max-min) * 230 ) / max );
+        }
+
+        return normalizedTwaps;
+    }
+
+    /*
+    * @dev Returns SVG of first point on graph.
+    */
+    function firstPointSVG(uint256 twap, uint256 normalizedTwap) internal pure returns (bytes memory) {
+
+        return abi.encodePacked(
+            '<circle cx="53" cy="',
+            Strings.toString(normalizedTwap),
+            '" r="5" fill="aqua" visibility=',
+            '"visible">',
+            '<set attributeName="visibility" ',
+            'from="visible" to="hidden" ',
+            'begin="0ms" dur="1000ms" />',
+            '</circle>',
+            '<text x="63" y="',
+            Strings.toString(normalizedTwap),
+            '" font-size="14">',
+            '<set attributeName="visibility" ',
+            'from="visible" to="hidden" ',
+            'begin="0ms" dur="1000ms" />',
+            // TODO what if twip has 10 leading 0's?
+            Strings.toString(twap), 
+            '</text>'
+        );
+    }
+    
+    /*
+    * @dev Returns SVG of remaining points on graph.
+    */
+    function remainingPointsSVG(uint256[] memory twaps, uint256[7] memory normalizedTwaps, uint16 xInterval) internal pure returns (bytes memory toReturn) {
```

Comment:
does this function return anywhere?

Reply:
Haha good point, it's not clear, but it returns the var 'toReturn' (see end of method signature)
I'm going to change it to explicitly return though, better readability